// manager/index.js
/**
 * Created by User on 22.11.2015.
 */
var gs = require('game_session');
var u = require('user');

var free_users = [];

var all_users = [];

var games = [];

exports.all = function(){
    return all_users;
};

exports.free = function(){
    return free_users;
};

exports.add_user = function(user){
    all_users.push(user);
};

exports.find_user = function(id){
    var res = all_users.filter(function(v){
        return v['id'] == id;
    });
    if(res.length > 0){
        return res[0];
    }

    return null;
};

exports.find_game = function(id){
    var res = games.filter(function(v){
        return v['id'] == id;
    });
    if(res.length > 0){
        return res[0];
    }

    return null;
};

exports.find_wait_game = function(){
    var res = games.filter(function(x){
        return x.users_id.length == 1;
    });
    if(res.length > 0){

        return res[0];
    }

    return null;
};

exports.start_new_game = function(game){
    games.push(game);
};

exports.finished_game = function(id){
    var game = this.find_game(id);
    free_users.push(this.find_user(game.users_id[0]));
    free_users.push(this.find_user(game.users_id[1]));
    games = games.filter(function(x){
        return x.id != id;
    });
};

exports.remove_wait_user = function(id){
    free_users = free_users.filter(function(x){
        return x.id != id;
    });
};

exports.remove_user = function(id){
    all_users = all_users.filter(function(x){
        return x.id != id;
    });
};
