// game_session/index.js
/**
 * Created by User on 22.11.2015.
 */
module.exports.create_new_game_session = function(id, user_id1){
    var game_session = {};
    game_session.id = id;
    game_session.users_id = [
        user_id1
    ];
    game_session.state = [
        [-1, -1, -1],
        [-1, -1, -1],
        [-1, -1, -1]
    ];
    game_session.current_user = user_id1;
    game_session.current_user_num = 0;
    game_session.finished = false;
    game_session.started = false;
    game_session.winner_count = 0;

    game_session.add_user = function(userid){
        this.users_id.push(userid);
        this.started = true;
    };

    game_session.get_winner = function(){
        return this.users_id[this.winner_value];
    }

    game_session.get_losser = function() {
        return this.users_id[1 - this.winner_value];
    }

    game_session.set_step = function(data){
        if(data.x == undefined || data.x == null || data.y == undefined || data.y == null){
            return false;
        }
        if(data.x >= 0 && data.x <3 && data.y >= 0 && data.y < 3){
            this.state[data.x][data.y] = this.current_user_num;
        }
        else{
            return false;
        }

        return true;
    }

    game_session.next_step = function(){
        var checked = this.check(this.state);
        if(this.finished){
            return false;
        }
        if(checked && this.winner_count == 2){
            return false;
        }
        if(checked){
            this.finished = true;
        }
        if(checked && this.current_user_num == 1){
            return false;
        }
        this.current_user_num = 1 - this.current_user_num;
        this.current_user = this.users_id[this.current_user_num];

        return true;
    }

    game_session.check = function(arr){
        this.winner_value = -1;
        this.winner_count = 0;
        var win_x = true;
        for (var i = 0; i < 3; i++) {
            win_x = arr[i][0] == arr[i][1] && arr[i][1] == arr[i][2] && arr[i][0] != -1;
            if(win_x){
                this.winner_count += 1;
                this.winner_value = arr[i][0];
            }
        }

        var win_y = true;
        for (i = 0; i < 3; i++) {
            win_y = arr[0][i] == arr[1][i] && arr[1][i] == arr[2][i] && arr[0][i] != -1;
            if (win_y) {
                this.winner_count += 1;
                this.winner_value = arr[i][0];
            }
        }

        var win_xy = (arr[0][0] == arr[1][1] && arr[1][1] == arr[2][2] && arr[0][0] != -1);

        var win_yx = (arr[0][2] == arr[1][1] && arr[1][1] == arr[2][0] && arr[0][2] != -1);

        if(win_xy){
            this.winner_count += 1;
            this.winner_value = arr[0][0];
        }

        if(win_yx){
            this.winner_count += 1;
            this.winner_value = arr[0][2];
        }

        return this.winner_count > 0;
    }

    return game_session;
};
